### ISSUE TITLE
Libtorch:how to create tensor from tensorRT fp16 cuda half type pointer?

### ISSUE THREAD

**Issue by Edwardmark** [pytorch/pytorch#37201]
how to create tensor from tensorRT fp16 half type pointer in libtorch?
I am working on a detection model. I change the backbone of it to tensorRT to do FP16 inference, and the detection code such as decode boxes and nms is done in libtorch and torchvisoin, so how to create fp16 tensor from tensorRT half type pointers?
The important code is to illustrate the issue:
```
// tensorRT code to get half type outpus
half_float::half* outputs[18];
doInference(*engine, data, outputs, 1);
// to get the final outputs with libtorch
vector<torch::Tensor> output;
//???? how to feed the date in outpus to output????
// get the result with libtorch method detect_trt->forward
 auto res = detect_trt->forward(output); 
```
Thanks in advance.

cc @yf225 @glaringlee

**Comment by gchanan** [comment]
CC @glaringlee.

**Comment by Edwardmark** [comment]
@glaringlee @gchanan 
I tried the following code and it worked, but I am wondering, what if the outputs is pointer of gpu data?
```
out = torch::from_blob(outputs[i], {1, num, dim, dim}, torch::kFloat16).to(device_used);
```

**Comment by glaringlee** [comment]
@Edwardmark
If the output ptr is from gpu, you need to pass a tensor options obj and set the device type to gpu device type, the default device type is CPU. Check here:
https://github.com/pytorch/pytorch/blob/9f0289743157b9c1e54ff47f7f11e3633f1ebf8f/aten/src/ATen/templates/Functions.h#L24

You code will be like
out = torch::from_blob(outputs[i], {1, num, dim, dim}, 
torch::Device(torch::kCUDA, 0).dtype(torch::kFloat16))

if you have on 1 graphic device, you can just do torch::Device(torch::kCUDA).dtype(torch::kFloat16) if you like.

if you have multiple graphic device. You can change 0 to your other graphic device index, 
but you have to make sure your input data ptr point to a memory chunk coming from the same device.

Let me know whether it works, thanks.

**Comment by Edwardmark** [comment]
@izdeby another question, 
in libtorch how to set all tensors same as 
'''
with torch.no_grad() 
```
in python?
I found cuda memory leak when I run the model. Any tools or clues about what could cause that?

**Comment by gchanan** [comment]
@Edwardmark there's something like torch::NoGradMode or torch::NoGradGuard.

**Comment by glaringlee** [comment]
@Edwardmark
An example using torch::NoGradGuard.
https://github.com/pytorch/pytorch/blob/50a1850d8d8e42bd3efe0661d309844bf0ac41dc/test/cpp/api/misc.cpp#L49

**Comment by Edwardmark** [comment]
@glaringlee  so we just add 
```
torch::NoGradGuard guard;
```
at the start of the torch code can make effect? Is that right?

**Comment by glaringlee** [comment]
@edwardmark yes, as long as that guard stay in the same scope as the rest of your code which you don't want grad calculation.

**Comment by Edwardmark** [comment]
@glaringlee Thanks very much.

**Comment by Edwardmark** [comment]
@glaringlee could you please help me about an indexing problem? Thanks. https://github.com/pytorch/pytorch/issues/37478
